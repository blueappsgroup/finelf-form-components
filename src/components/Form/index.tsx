import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
