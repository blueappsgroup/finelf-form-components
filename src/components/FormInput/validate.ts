import {
  isEmpty,
  isNotValidLength,
  isNotValidPhoneNumber,
  isNotValidEmail,
  isNotLetter,
} from './validateHelpers'

export const validateText: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidLength(value)) {
    return 'Pole musi zawierać między 2 - 100 znaków'
  }

  if (isNotLetter(value)) {
    return 'Podane dane są nieprawidłowe'
  }
}

export const validateEmail: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany email jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidEmail(value)) {
    return customErrorMsg
  }
}

export const validateTelNumber: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany numer telefonu jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidPhoneNumber(value)) {
    return customErrorMsg
  }
}

export const validateCheckbox: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => (required && !value && 'required') || undefined
