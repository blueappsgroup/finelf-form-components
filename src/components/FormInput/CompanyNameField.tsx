import React, { FC } from 'react'

import { TextFieldWithConditonProps } from '../../types'
import BaseField from './base'
import { validateText } from './validate'
import CustomFieldWithCondition from './CustomFieldWithCondition'

const CompanyNameField: FC<TextFieldWithConditonProps> = ({
  id,
  label = 'Nazwa pracodawcy',
  placeholder = 'Nazwa pracodawcy',
  name = 'company_name',
  required = false,
  validate = validateText,
  showError = true,
  requiredCondition,
  visibleCondition = {
    // eslint-disable-next-line @typescript-eslint/camelcase
    employment_type: [
      'fixed',
      'temporary',
      'fixed_partial',
      'government',
      'agriculture',
      'service_contract',
    ],
  },
}) => (
  <CustomFieldWithCondition
    id={id}
    label={label}
    placeholder={placeholder}
    component={BaseField}
    required={required}
    showError={showError}
    name={name}
    validate={validate}
    requiredCondition={requiredCondition}
    visibleCondition={visibleCondition}
  />
)

CompanyNameField.defaultProps = {
  name: 'company_name',
}

export default CompanyNameField
