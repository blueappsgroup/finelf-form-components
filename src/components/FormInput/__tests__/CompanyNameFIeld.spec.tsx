import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import CompanyNameField from '../CompanyNameField'
import { validateText } from '../validate'

describe('<CompanyNameField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    inputId = 'testId',
    inputName = 'company_name',
    required = false,
    validate,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <CompanyNameField
          id={inputId}
          name={inputName}
          validate={validate}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'CompanyName' })
    const { container } = wrapper

    expect(wrapper.getByText('CompanyName')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, {
        target: { value: 'Name' },
      })
    })

    expect(input.value).toBe('Name')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, {
        target: { value: 'Name' },
      })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'Name' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe('Name')
    expect(input.value).toBe('Name')
  })

  it('matches snapshot with error', async () => {
    const wrapper = setupWrapper({
      label: 'CompanyName',
      showError: true,
      required: true,
      error: 'true',
    })
    const { container, input } = wrapper

    await act(async () => {
      fireEvent.focus(input, {
        target: { value: 'Name' },
      })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'Name' } })
    })

    expect(wrapper.getByText('To pole jest wymagane')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('input field with custom validation', async () => {
    const inputValue = '3232'

    const wrapper = setupWrapper({
      validate: validateText,
      showError: true,
    })

    const { input, container } = wrapper

    await act(async () => {
      fireEvent.change(input, { target: { value: inputValue } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: inputValue } })
    })

    expect(container).toMatchSnapshot()
    expect(wrapper.getByText('Podane dane są nieprawidłowe')).toBeTruthy()
  })
})
