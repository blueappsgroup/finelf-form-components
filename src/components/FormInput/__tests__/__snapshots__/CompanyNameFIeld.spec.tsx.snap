// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompanyNameField /> input field with custom validation 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="company_name"
      >
        Nazwa pracodawcy
      </label>
      <div
        class="sc-iJuUWI cFUPqY"
      >
        <input
          class="sc-fodVxV gZGAuo"
          id="testId"
          label="Nazwa pracodawcy"
          name="company_name"
          placeholder="Nazwa pracodawcy"
          value="3232"
        />
      </div>
      <span
        class="sc-hBEYos kIOtFs"
      >
        Podane dane są nieprawidłowe
      </span>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      disabled=""
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`<CompanyNameField /> matches snapshot 1`] = `
<input
  class="sc-fodVxV kubVCF"
  id="testId"
  label="Nazwa pracodawcy"
  name="company_name"
  placeholder="Nazwa pracodawcy"
  value=""
/>
`;

exports[`<CompanyNameField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="company_name"
      >
        CompanyName*
      </label>
      <div
        class="sc-iJuUWI cFUPqY"
      >
        <input
          class="sc-fodVxV gZGAuo"
          id="testId"
          label="CompanyName"
          name="company_name"
          placeholder="Nazwa pracodawcy*"
          required=""
          value=""
        />
      </div>
      <span
        class="sc-hBEYos kIOtFs"
      >
        To pole jest wymagane
      </span>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      disabled=""
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`<CompanyNameField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="company_name"
      >
        CompanyName
      </label>
      <div
        class="sc-iJuUWI cFUPqY"
      >
        <input
          class="sc-fodVxV kubVCF"
          id="testId"
          label="CompanyName"
          name="company_name"
          placeholder="Nazwa pracodawcy"
          value=""
        />
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;
